File: 골드 5 (6문제)/9251_LCS.c
// 설명 링크 : https://moon-gd.tistory.com/96

#define _CRT_SECURE_NO_WARNINGS

#include <stdio.h>
#include <string.h>
#define MAX(a, b) a > b ? a : b
#define MAX_LENGTH 1000

char A[MAX_LENGTH + 1];
char B[MAX_LENGTH + 1];

int dp[MAX_LENGTH + 1][MAX_LENGTH + 1];

int main() {
	// 문자열 입력
	scanf("%s", &A);
	scanf("%s", &B);
	
	// 길이 확인
	int len_A = strlen(A);
	int len_B = strlen(B);

	// 결과 출력 : 빈 문자열이 있는 경우 0 출력 후 종료
	if (len_A == 0 || len_B == 0) {
		printf("0");
		return 0;
	}

	for (int i = 1; i <= len_B; i++) {
		for (int j = 1; j <= len_A; j++) {
			if (B[i - 1] == A[j - 1]) {
				dp[j][i] = dp[j - 1][i - 1] + 1;
			}
			else {
				dp[j][i] = MAX(dp[j - 1][i], dp[j][i - 1]);
			}
		}
	}

	printf("%d", dp[len_A][len_B]);

	return 0;
}

File: 골드 5 (6문제)/9251_LCS.py
# 설명 링크 : https://moon-gd.tistory.com/96
import sys
q_input = sys.stdin.readline  # 빠른 입력

# 문자열 입력
A = q_input().rstrip()
B = q_input().rstrip()

# 문자열 길이 계산
len_A, len_B = len(A), len(B)

if len_A == 0 or len_B == 0:
    print(0)
else:
    # DP 배열 생성
    dp = []
    for _ in range(len_B + 1):
        dp.append([0] * (len_A + 1))

    # DP 배열 계산
    for i in range(1, len_B + 1):
        for j in range(1, len_A + 1):
            if A[j - 1] == B[i - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    # 결과 출력
    print(dp[len_B][len_A])
